File: python/395.py
#!/usr/bin/env python3
"""Project Euler Problem 395 - Pythagorean Tree

Find minimum area of rectangle parallel to initial square containing entire tree.

Solution iteratively finds extreme points in each direction using high-precision
arithmetic (mpmath) to avoid rounding errors in the last digit.
"""

from mpmath import mp, mpf, sqrt, fabs

mp.dps = 50  # 50 decimal places of precision

class Point:
    def __init__(self, x, y):
        self.x = mpf(x)
        self.y = mpf(y)

    def add(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def subtract(self, other):
        return Point(self.x - other.x, self.y - other.y)

    def complex_multiply(self, other):
        """Multiply as complex numbers"""
        return Point(
            self.x * other.x - self.y * other.y,
            self.x * other.y + self.y * other.x
        )

    def dist(self, other):
        dx = self.x - other.x
        dy = self.y - other.y
        return sqrt(dx * dx + dy * dy)


def solve():
    A, B, C = mpf(3), mpf(4), mpf(5)
    L = sqrt(2) / (1 - B / C)

    def find_extreme(get_coord):
        """Find extreme point in direction given by get_coord function"""
        segments = [((Point(0, 0), Point(1, 0)))]
        extremity = mpf('-inf')

        while True:
            new_segments = []

            for start, end in segments:
                diff = end.subtract(start)

                # Four transformation points for the new squares
                transforms = [
                    Point(-A * B / (C * C), B * B / (C * C)),
                    Point((B - A) * B / (C * C), (B + A) * B / (C * C)),
                    Point((B + A) * B / (C * C), (B + A) * A / (C * C)),
                    Point(1 + A * B / (C * C), A * A / (C * C))
                ]

                new_points = [start.add(diff.complex_multiply(t)) for t in transforms]

                new_segments.append((new_points[0], new_points[1]))
                new_segments.append((new_points[2], new_points[3]))

            # Find new extreme point
            all_points = {}
            for seg in new_segments:
                all_points[get_coord(seg[0])] = seg
                all_points[get_coord(seg[1])] = seg

            new_extremity = max(all_points.keys())

            if fabs(extremity - new_extremity) < mpf('1e-30'):
                break

            extremity = new_extremity

            # Prune segments that can't contribute further
            segments_set = set()
            for coord, seg in all_points.items():
                seg_len = seg[0].dist(seg[1])
                if coord > new_extremity - L * seg_len:
                    segments_set.add(seg)

            segments = list(segments_set)

        return extremity

    # Find extremes in all four directions
    min_x = -find_extreme(lambda p: -p.x)
    min_y = -find_extreme(lambda p: -p.y)
    max_x = find_extreme(lambda p: p.x)
    max_y = find_extreme(lambda p: p.y)

    area = (max_x - min_x) * (max_y - min_y)
    # Format to 10 decimal places
    return mp.nstr(area, 12, strip_zeros=False)

if __name__ == "__main__":
    print(solve())

File: c/395.c
/*
 * Project Euler Problem 395: Pythagorean Tree
 *
 * Find minimum bounding rectangle area for the infinite Pythagorean tree.
 * Uses iterative refinement of extreme points with high precision (long double).
 */
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SEGS 100000

typedef struct { long double x, y; } Point;

static Point point_add(Point a, Point b) {
    return (Point){a.x + b.x, a.y + b.y};
}

static Point point_sub(Point a, Point b) {
    return (Point){a.x - b.x, a.y - b.y};
}

static Point complex_mul(Point a, Point b) {
    return (Point){a.x * b.x - a.y * b.y, a.x * b.y + a.y * b.x};
}

static long double point_dist(Point a, Point b) {
    long double dx = a.x - b.x, dy = a.y - b.y;
    return sqrtl(dx * dx + dy * dy);
}

typedef struct { Point start, end; } Segment;

static long double find_extreme(int dir) {
    /* dir: 0=max_x, 1=max_y, 2=min_x(negated), 3=min_y(negated) */
    long double A = 3.0L, B = 4.0L, C = 5.0L;
    long double L_bound = sqrtl(2.0L) / (1.0L - B / C);

    Segment *segs = malloc(MAX_SEGS * sizeof(Segment));
    Segment *new_segs = malloc(MAX_SEGS * sizeof(Segment));
    int nsegs = 1;
    segs[0] = (Segment){{0, 0}, {1, 0}};

    long double extremity = -1e30L;

    for (int iter = 0; iter < 100; iter++) {
        int nnew = 0;

        Point transforms[4] = {
            {-A * B / (C * C), B * B / (C * C)},
            {(B - A) * B / (C * C), (B + A) * B / (C * C)},
            {(B + A) * B / (C * C), (B + A) * A / (C * C)},
            {1.0L + A * B / (C * C), A * A / (C * C)}
        };

        for (int i = 0; i < nsegs && nnew + 2 <= MAX_SEGS; i++) {
            Point diff = point_sub(segs[i].end, segs[i].start);
            Point pts[4];
            for (int j = 0; j < 4; j++) {
                pts[j] = point_add(segs[i].start, complex_mul(diff, transforms[j]));
            }
            new_segs[nnew++] = (Segment){pts[0], pts[1]};
            new_segs[nnew++] = (Segment){pts[2], pts[3]};
        }

        /* Find new extreme */
        long double new_extremity = -1e30L;
        for (int i = 0; i < nnew; i++) {
            long double v1, v2;
            switch (dir) {
                case 0: v1 = new_segs[i].start.x; v2 = new_segs[i].end.x; break;
                case 1: v1 = new_segs[i].start.y; v2 = new_segs[i].end.y; break;
                case 2: v1 = -new_segs[i].start.x; v2 = -new_segs[i].end.x; break;
                case 3: v1 = -new_segs[i].start.y; v2 = -new_segs[i].end.y; break;
                default: v1 = v2 = 0; break;
            }
            if (v1 > new_extremity) new_extremity = v1;
            if (v2 > new_extremity) new_extremity = v2;
        }

        if (fabsl(extremity - new_extremity) < 1e-25L) break;
        extremity = new_extremity;

        /* Prune segments that can't contribute */
        int npruned = 0;
        for (int i = 0; i < nnew; i++) {
            long double seg_len = point_dist(new_segs[i].start, new_segs[i].end);
            long double v1, v2;
            switch (dir) {
                case 0: v1 = new_segs[i].start.x; v2 = new_segs[i].end.x; break;
                case 1: v1 = new_segs[i].start.y; v2 = new_segs[i].end.y; break;
                case 2: v1 = -new_segs[i].start.x; v2 = -new_segs[i].end.x; break;
                case 3: v1 = -new_segs[i].start.y; v2 = -new_segs[i].end.y; break;
                default: v1 = v2 = 0; break;
            }
            long double max_v = (v1 > v2) ? v1 : v2;
            if (max_v > new_extremity - L_bound * seg_len) {
                if (npruned < MAX_SEGS) {
                    segs[npruned++] = new_segs[i];
                }
            }
        }
        nsegs = npruned;
    }

    free(segs);
    free(new_segs);
    return extremity;
}

int main(void) {
    long double max_x = find_extreme(0);
    long double max_y = find_extreme(1);
    long double min_x_neg = find_extreme(2); /* -min_x */
    long double min_y_neg = find_extreme(3); /* -min_y */

    long double width = max_x + min_x_neg;
    long double height = max_y + min_y_neg;
    long double area = width * height;

    /* Format to match expected output */
    printf("%.10Lf\n", area);
    return 0;
}
